File: Cocoa/AppFrameworks/UIKit/ScrollViewManager/YJUIScrollViewManager.m
//
//  YJUIScrollViewManager.m
//  ScrollViewManager
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2016/12/22.
//  Copyright © 2016年 YJCocoa. All rights reserved.
//

#import "YJUIScrollViewManager.h"
#import "YJNSAspectOrientProgramming.h"
#import "UIView+YJUIViewGeometry.h"

@interface YJUIScrollViewManager () <UIScrollViewDelegate> {
    BOOL _verticalScrollEnable;   ///< 垂直开关
    BOOL _horizontalScrollEnable; ///< 水平开关
    CGPoint _contentOffset; ///< 起点
}

@property (nonatomic, strong) YJNSAspectOrientProgramming *aopDelegate; ///< 切面代理
@property (nonatomic) YJUIScrollViewScroll verticalScroll;   ///< 垂直滚动
@property (nonatomic) YJUIScrollViewScroll horizontalScroll; ///< 水平滚动

@end

@implementation YJUIScrollViewManager

- (instancetype)initWithScrollView:(UIScrollView *)scrollView {
    self = [self init];
    if (self) {
        _scrollView = scrollView;
        self.didSpacing = 30;
        self.edgeInset = UIEdgeInsetsZero;
        self.endInset = UIEdgeInsetsZero;
        _scrollView.delegate = self;
    }
    return self;
}

- (void)addScrollViewAOPDelegate:(id<UIScrollViewDelegate>)delegate {
    if (!_aopDelegate) {
        _aopDelegate = [[YJNSAspectOrientProgramming alloc] init];
        [_aopDelegate addTarget:self];
    }
    [_aopDelegate addTarget:delegate];
    self.scrollView.delegate = (id<UIScrollViewDelegate>)self.aopDelegate;
}

#pragma mark - UIScrollViewDelegate
- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView {
    _contentOffset = scrollView.contentOffset;
    if (_verticalScrollEnable)   self.verticalScroll = YJUIScrollViewScrollNone;
    if (_horizontalScrollEnable) self.horizontalScroll = YJUIScrollViewScrollNone;
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    if (_verticalScrollEnable)   [self scrollViewDidVerticalScroll:scrollView];
    if (_horizontalScrollEnable) [self scrollViewDidHorizontalScroll:scrollView];
}

#pragma mark - private
- (void)scrollViewDidVerticalScroll:(UIScrollView *)scrollView {
    CGFloat y = scrollView.contentOffset.y;
    if (y <= self.endInset.top) {
        self.verticalScroll = YJUIScrollViewScrollEndTop;
    } else if (y <= self.edgeInset.top) {
        self.verticalScroll = YJUIScrollViewScrollEdgeTop;
    } else {
        CGFloat scrollHeight = scrollView.heightFrame > scrollView.contentSize.height ? scrollView.heightFrame : scrollView.contentSize.height;
        CGFloat bottom = y + scrollView.heightFrame;
        if (bottom >= scrollHeight - self.endInset.bottom) {
            self.verticalScroll = YJUIScrollViewScrollEndBottom;
        } else if (bottom >= scrollHeight - self.edgeInset.bottom) {
            self.verticalScroll = YJUIScrollViewScrollEdgeBottom;
        } else {
            CGFloat spacing = y - _contentOffset.y;
            if (spacing <= -self.didSpacing) {
                self.verticalScroll = YJUIScrollViewScrollDidTop;
                _contentOffset.y = y;
            } else if (spacing >= self.didSpacing) {
                self.verticalScroll = YJUIScrollViewScrollDidBottom;
                _contentOffset.y = y;
            }
        }
    }
}

- (void)scrollViewDidHorizontalScroll:(UIScrollView *)scrollView {
    CGFloat x = scrollView.contentOffset.x;
    if (x <= self.endInset.left) {
        self.horizontalScroll = YJUIScrollViewScrollEndLeft;
    } else if (x <= self.edgeInset.left) {
        self.horizontalScroll = YJUIScrollViewScrollEdgeLeft;
    } else {
        CGFloat scrollWidth = scrollView.widthFrame > scrollView.contentSize.width ? scrollView.widthFrame : scrollView.contentSize.width;
        CGFloat right = x + scrollView.widthFrame;
        if (right >= scrollWidth - self.endInset.right) {
            self.horizontalScroll = YJUIScrollViewScrollEndRight;
        } else if (right >= scrollWidth - self.edgeInset.right) {
            self.horizontalScroll = YJUIScrollViewScrollEdgeRight;
        } else {
            CGFloat spacing = x - _contentOffset.x;
            if (spacing <= -self.didSpacing) {
                self.horizontalScroll = YJUIScrollViewScrollDidLeft;
                _contentOffset.x = x;
            } else if (spacing >= self.didSpacing) {
                self.horizontalScroll = YJUIScrollViewScrollDidRight;
                _contentOffset.x = x;
            }
        }
    }
}

#pragma mark - getter & setter
- (void)setDelegate:(id<YJUIScrollViewManagerDelegate>)delegate {
    _delegate = delegate;
    _verticalScrollEnable = [delegate respondsToSelector:@selector(scrollViewManager:didVerticalScroll:)];
    _horizontalScrollEnable = [delegate respondsToSelector:@selector(scrollViewManager:didHorizontalScroll:)];
}

- (void)setVerticalScroll:(YJUIScrollViewScroll)verticalScroll {
    if (_verticalScroll == YJUIScrollViewScrollNone || _verticalScroll != verticalScroll) {
        _verticalScroll = verticalScroll;
        [self.delegate scrollViewManager:self didVerticalScroll:verticalScroll];
    }
}

- (void)setHorizontalScroll:(YJUIScrollViewScroll)horizontalScroll {
    if (_horizontalScroll == YJUIScrollViewScrollNone || _horizontalScroll != horizontalScroll) {
        _horizontalScroll = horizontalScroll;
        [self.delegate scrollViewManager:self didHorizontalScroll:horizontalScroll];
    }
}

@end

File: YJCocoa/Classes/AppFrameworks/UIKit/ScrollView/YJUIScrollViewManager.swift
//
//  YJUIScrollViewManager.swift
//  YJCocoa
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2019/5/23.
//  Copyright © 2016-现在 YJCocoa. All rights reserved.
//

import UIKit

/// UIScrollView滚动
public enum YJUIScrollViewScroll: Int {
    // 用户触摸，将要滚动
    case none
    
    /// 滚动到顶部
    case endTop
    /// 滚动到边缘顶部
    case edgeTop
    /// 向上滚动
    case didTop
    
    /// 向下滚动
    case didBottom
    /// 滚动到边缘底部
    case edgeBottom
    /// 滚动到底部
    case endBottom
    
    /// 滚动到最左
    case endLeft
    /// 滚动到边缘左
    case edgeLeft
    /// 向左滚动
    case didLeft
    
    /// 向右滚动
    case didRight
    /// 滚动到边缘右
    case edgeRight
    /// 滚动到最右
    case endRight
}

/// UIScrollView管理器
open class YJUIScrollViewManager: NSObject {
    
    public typealias DidScrollClosure = (_ scroll: YJUIScrollViewScroll) -> Void
    
    /// 已经滚动间隔,控制.Did...枚举提示间隔
    public var didSpacing: CGFloat = 30
    /// 边缘距离,控制.Edge...枚举提示区域,需edgeInset内值>endInset内值
    public var edgeInset = UIEdgeInsets.zero
    /// 边界距离,控制.End...枚举提示区域
    public var endInset = UIEdgeInsets.zero
    
    /// 垂直滚动回调
    public var didVerticalScrollClosure: DidScrollClosure?
    /// 水平滚动回调
    public var didHorizontalScrollClosure: DidScrollClosure?
    
    private var contentOffset = CGPoint.zero
    private var verticalScroll = YJUIScrollViewScroll.none {
        willSet {
            if newValue != self.verticalScroll || newValue == .none  {
                self.didVerticalScrollClosure!(newValue)
            }
        }
    }
    private var horizontalScroll = YJUIScrollViewScroll.none {
        willSet {
            if newValue != self.horizontalScroll || newValue == .none  {
                self.didHorizontalScrollClosure!(newValue)
            }
        }
    }
    
}

extension YJUIScrollViewManager: UIScrollViewDelegate {
    
    public func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        self.contentOffset = scrollView.contentOffset
        if self.didVerticalScrollClosure != nil {
            self.verticalScroll = .none
        }
        if self.didHorizontalScrollClosure != nil {
            self.horizontalScroll = .none
        }
    }
    
    public func scrollViewDidScroll(_ scrollView: UIScrollView) {
        if self.didVerticalScrollClosure != nil {
            self.scrollViewDidVerticalScroll(scrollView)
        }
        if self.didHorizontalScrollClosure != nil {
            self.scrollViewDidHorizontalScroll(scrollView)
        }
    }
    
    func scrollViewDidVerticalScroll(_ scrollView: UIScrollView) {
        let y = scrollView.contentOffset.y;
        let maxY = scrollView.contentSize.height - scrollView.frameHeight
        guard maxY > 0 else {
            return
        }
        let spacing = y - self.contentOffset.y;
        if (y <= self.endInset.top) {
            self.verticalScroll = .endTop;
        } else if (y <= self.edgeInset.top) {
            self.verticalScroll = .edgeTop;
        } else if (y >= maxY - self.endInset.bottom) {
            self.verticalScroll = .endBottom;
        } else if (y >= maxY - self.edgeInset.bottom) {
            self.verticalScroll = .edgeBottom;
        } else if spacing <= -self.didSpacing {
            self.verticalScroll = .didTop;
        } else if spacing >= self.didSpacing {
            self.verticalScroll = .didBottom;
        } else {
            return
        }
        self.contentOffset.y = y;
    }
    
    func scrollViewDidHorizontalScroll(_ scrollView: UIScrollView) {
        let x = scrollView.contentOffset.x;
        let maxX = scrollView.contentSize.width - scrollView.frameWidth
        guard maxX > 0 else {
            return
        }
        let spacing = x - self.contentOffset.x;
        if (x <= self.endInset.left) {
            self.horizontalScroll = .endLeft;
        } else if (x <= self.edgeInset.left) {
            self.horizontalScroll = .edgeLeft;
        } else if (x >= maxX - self.endInset.right) {
            self.horizontalScroll = .endRight;
        } else if (x >= maxX - self.edgeInset.right) {
            self.horizontalScroll = .edgeRight;
        } else if spacing <= -self.didSpacing {
            self.horizontalScroll = .didLeft;
        } else if spacing >= self.didSpacing {
            self.horizontalScroll = .didRight;
        } else {
            return
        }
        self.contentOffset.x = x;
    }
    
}

